File: essentials/chapter_6/agent_memory_retrieval.py
"""Agent Selective Memory Retrieval - LangChain with BM25."""

import os
import re
import sys
from typing import Dict, List, Optional

REPO_ROOT = os.path.abspath(os.path.join(os.path.dirname(__file__), "..", "..", ".."))
sys.path.append(REPO_ROOT)

if any(arg in {"-h", "--help"} for arg in sys.argv[1:]):
    from shared.utils import print_cli_help

    print_cli_help(sys.argv[0])
    sys.exit(0)

from langchain_core.documents import Document
from langchain_core.messages import HumanMessage
from langchain_core.messages.utils import count_tokens_approximately
from langchain_community.retrievers import BM25Retriever


from essentials.chapter_5.agent_structured_output import STRUCTURED_TEMPLATE, LangChainLLMManager as Chapter5StructuredManager
from shared.essentials.utils import interactive_cli, parse_structured_json_response


class LangChainLLMManager(Chapter5StructuredManager):
    """Chapter 6 manager that retrieves only relevant memory before prompting the LLM."""

    def __init__(self, memory_enabled: bool = True, retrieval_k: int = 4):
        # Keep Chapter 5's inherited full-history chain disabled because this manager
        # builds its own prompt with retrieved snippets instead of replaying all turns.
        super().__init__(memory_enabled=False)
        # Store Chapter 6 retrieval-memory behavior separately for clarity.
        self.retrieval_memory_enabled = memory_enabled
        self.retrieval_k = max(1, retrieval_k)
        self.framework = "LangChain Memory+Retrieval"

    @staticmethod
    def _estimate_tokens(text: str) -> int:
        if not text:
            return 0
        return int(count_tokens_approximately([HumanMessage(content=text)]))

    @staticmethod
    def _tokenize(text: str) -> set:
        stop_words = {
            "a", "an", "and", "are", "as", "at", "be", "but", "by", "for", "from", "how",
            "i", "in", "is", "it", "of", "on", "or", "that", "the", "this", "to", "was",
            "we", "what", "when", "where", "which", "who", "why", "with", "you",
        }
        tokens = set(re.findall(r"[a-zA-Z0-9_]+", str(text).lower()))
        return {token for token in tokens if len(token) > 2 and token not in stop_words}

    def _overlap_score(self, query_tokens: set, content: str) -> int:
        content_tokens = self._tokenize(content)
        if not query_tokens or not content_tokens:
            return 0
        return len(query_tokens.intersection(content_tokens))

    def _select_retrieved_messages(self, topic: str, messages: List) -> List[Dict[str, str]]:
        query_tokens = self._tokenize(topic)
        if not query_tokens:
            return []

        documents = []
        for idx, msg in enumerate(messages):
            content = str(getattr(msg, "content", "") or "")
            if not content:
                continue
            role = getattr(msg, "type", "unknown")
            documents.append(Document(page_content=content, metadata={"idx": idx, "role": role}))

        if documents:
            retriever = BM25Retriever.from_documents(documents, k=self.retrieval_k, include_score=True)
            retrieved_docs = retriever.invoke(topic)
            strong_docs = []
            for doc in retrieved_docs:
                bm25 = float(doc.metadata.get("bm25Score", 0))
                if bm25 <= 0:
                    continue
                overlap = self._overlap_score(query_tokens, str(doc.page_content))
                overlap_ratio = overlap / len(query_tokens)
                if overlap >= 2 or overlap_ratio >= 0.4:
                    strong_docs.append(doc)

            if strong_docs:
                top = sorted(strong_docs, key=lambda doc: doc.metadata.get("idx", 0))
                return [
                    {
                        "role": str(doc.metadata.get("role", "unknown")),
                        "content": str(doc.page_content),
                        "relevance_score": float(doc.metadata.get("bm25Score", 0)),
                    }
                    for doc in top
                ]

        fallback = []
        for idx, msg in enumerate(messages):
            content = str(getattr(msg, "content", "") or "")
            if not content:
                continue
            overlap = self._overlap_score(query_tokens, content)
            overlap_ratio = overlap / len(query_tokens)
            if overlap >= 2 or overlap_ratio >= 0.4:
                fallback.append((overlap, idx, msg))

        if not fallback:
            return []

        fallback.sort(key=lambda item: (-item[0], item[1]))
        top = sorted(fallback[: self.retrieval_k], key=lambda item: item[1])
        return [
            {
                "role": getattr(msg, "type", "unknown"),
                "content": str(getattr(msg, "content", "")),
                "relevance_score": score,
            }
            for score, _, msg in top
        ]

    def ask_question(
        self,
        topic: str,
        provider: Optional[str] = None,
        template: str = STRUCTURED_TEMPLATE,
        max_tokens: int = 1000,
        temperature: float = 0.7,
        session_id: str = "default",
    ) -> Dict:
        effective_template = STRUCTURED_TEMPLATE if template == "{topic}" else template
        provider = self.resolve_model_identifier(provider)
        base_prompt = effective_template.format(topic=topic)

        if not provider:
            return {
                "success": False,
                "error": "No providers available",
                "provider": "none",
                "model": "none",
                "prompt": base_prompt,
                "response": None,
            }

        messages = self._get_history(session_id).messages if self.retrieval_memory_enabled else []
        retrieved = self._select_retrieved_messages(topic, messages) if self.retrieval_memory_enabled else []

        retrieved_context = "\n".join(f"[{item['role']}] {item['content']}" for item in retrieved)
        retrieval_augmented_topic = (
            f"Relevant memory snippets:\n{retrieved_context}\n\nCurrent user topic: {topic}"
            if retrieved_context
            else topic
        )
        retrieval_prompt = effective_template.format(topic=retrieval_augmented_topic)

        full_history_context = "\n".join(
            f"[{getattr(msg, 'type', 'unknown')}] {getattr(msg, 'content', '')}" for msg in messages
        )
        prompt_without_retrieval = (
            f"{full_history_context}\n\nCurrent user topic: {topic}" if full_history_context else topic
        )

        tokens_with_retrieval = self._estimate_tokens(retrieval_prompt)
        tokens_without_retrieval = self._estimate_tokens(prompt_without_retrieval)
        estimated_saved = max(0, tokens_without_retrieval - tokens_with_retrieval)
        reduction_percent = (
            round((estimated_saved / tokens_without_retrieval) * 100, 2) if tokens_without_retrieval else 0.0
        )

        model_config = self.resolve_model_config(provider)

        try:
            model = self._create_model(provider, temperature=temperature, max_tokens=max_tokens)
            result = model.invoke(self._build_messages(retrieval_prompt))
            raw_response = self._extract_text(provider, result)
            response_metadata = getattr(result, "response_metadata", None)
            usage_metadata = getattr(result, "usage_metadata", None)
            token_usage = self._extract_token_usage(response_metadata, usage_metadata)

            result_payload = {
                "provider": model_config.provider,
                "model": model_config.model,
                "model_identifier": model_config.name,
                "session_id": session_id,
                "temperature": temperature,
                "max_tokens": max_tokens,
                "response_metadata": response_metadata,
                "usage_metadata": usage_metadata,
            }

            try:
                parsed = parse_structured_json_response(raw_response)
            except Exception:
                parsed = {
                    "answer": raw_response,
                    "summary": "Model returned plain-text output instead of strict JSON.",
                    "keywords": [],
                    "distilled": raw_response,
                    "metadata": {
                        "confidence": "low",
                        "notes": "Structured parser fallback applied for non-JSON response.",
                    },
                }

            parsed["metadata"] = {
                **(parsed.get("metadata") or {}),
                **self._build_metadata(result_payload, raw_response),
                "retrieval": {
                    "history_messages_available": len(messages),
                    "retrieved_messages_count": len(retrieved),
                    "retrieved_messages": retrieved,
                    "tokens_with_memory_retrieval": tokens_with_retrieval,
                    "tokens_without_memory_retrieval": tokens_without_retrieval,
                    "estimated_tokens_saved": estimated_saved,
                    "estimated_token_reduction_percent": reduction_percent,
                },
            }

            if self.retrieval_memory_enabled:
                history = self._get_history(session_id)
                history.add_user_message(topic)
                history.add_ai_message(raw_response)

            payload = {
                "success": True,
                "provider": model_config.provider,
                "model": model_config.model,
                "model_identifier": model_config.name,
                "prompt": retrieval_prompt,
                "response": parsed,
                "raw_answer": parsed.get("answer", raw_response),
                "temperature": temperature,
                "max_tokens": max_tokens,
                "session_id": session_id,
                "response_metadata": response_metadata,
                "usage_metadata": usage_metadata,
                "token_usage": token_usage,
            }
            return {k: v for k, v in payload.items() if v is not None}
        except Exception as exc:
            return {
                "success": False,
                "provider": model_config.provider,
                "model": model_config.model,
                "model_identifier": model_config.name,
                "prompt": retrieval_prompt,
                "error": str(exc),
                "response": None,
                "temperature": temperature,
                "max_tokens": max_tokens,
                "session_id": session_id,
            }


def main():
    args = sys.argv[1:]
    if "web" in args:
        from shared.essentials.web import run_web_server

        run_web_server(lambda: LangChainLLMManager(memory_enabled=True))
    else:
        interactive_cli(LangChainLLMManager(memory_enabled=True))


if __name__ == "__main__":
    main()

File: essentials/chapter_6/agent_memory_retrieval.ts
/**
 * Agent Selective Memory Retrieval - LangChain TypeScript with BM25.
 */

import { LangChainLLMManager as Chapter5StructuredLangChainManager, STRUCTURED_TEMPLATE } from '../chapter_5/agent_structured_output.ts';
import { interactiveCli, parseStructuredJsonResponse, printCliHelp } from '../../../shared/essentials/utils.ts';
import { Document } from '@langchain/core/documents';
import { BM25Retriever } from '@langchain/community/retrievers/bm25';
import { getEncoding } from 'js-tiktoken';

class LangChainLLMManager extends Chapter5StructuredLangChainManager {
    constructor(memoryEnabled = true, retrievalK = 4) {
        // Disable inherited full-history replay chain. Chapter 6 builds retrieval prompts directly.
        super(false);
        this.framework = 'LangChain Memory+Retrieval TypeScript';
        this.retrievalMemoryEnabled = memoryEnabled;
        this.retrievalK = Math.max(1, retrievalK);
        // Provider-agnostic tokenizer baseline (GPT-2 BPE), without model/provider mapping.
        this.tokenizer = getEncoding('gpt2');
    }

    _estimateTokens(text) {
        if (!text) return 0;
        return this.tokenizer.encode(String(text)).length;
    }

    _tokenize(text) {
        const stopWords = new Set([
            'a', 'an', 'and', 'are', 'as', 'at', 'be', 'but', 'by', 'for', 'from', 'how',
            'i', 'in', 'is', 'it', 'of', 'on', 'or', 'that', 'the', 'this', 'to', 'was',
            'we', 'what', 'when', 'where', 'which', 'who', 'why', 'with', 'you',
        ]);
        const tokens = String(text || '').toLowerCase().match(/[a-z0-9_]+/g) || [];
        return new Set(tokens.filter((token) => token.length > 2 && !stopWords.has(token)));
    }

    _overlapScore(queryTokens, content) {
        const contentTokens = this._tokenize(content);
        if (queryTokens.size === 0 || contentTokens.size === 0) return 0;
        let score = 0;
        for (const token of queryTokens) {
            if (contentTokens.has(token)) score += 1;
        }
        return score;
    }

    async _selectRetrievedMessages(topic, messages) {
        const queryTokens = this._tokenize(topic);
        if (queryTokens.size === 0) return [];

        const docs = messages
            .map((msg, idx) => {
                const content = String(msg?.content ?? '');
                if (!content) return null;
                return new Document({
                    pageContent: content,
                    metadata: {
                        idx,
                        role: msg?._getType?.() ?? msg?.getType?.() ?? msg?.type ?? msg?.role ?? 'unknown',
                    },
                });
            })
            .filter(Boolean);

        if (docs.length > 0) {
            const retriever = BM25Retriever.fromDocuments(docs, { k: this.retrievalK, includeScore: true });
            const retrievedDocs = await retriever.invoke(topic);
            const strongDocs = retrievedDocs.filter((doc) => {
                const bm25 = Number(doc?.metadata?.bm25Score ?? 0);
                if (bm25 <= 0) return false;
                const overlap = this._overlapScore(queryTokens, String(doc?.pageContent ?? ''));
                const overlapRatio = overlap / queryTokens.size;
                return overlap >= 2 || overlapRatio >= 0.4;
            });

            if (strongDocs.length > 0) {
                return strongDocs
                    .sort((a, b) => (a?.metadata?.idx ?? 0) - (b?.metadata?.idx ?? 0))
                    .map((doc) => ({
                        role: doc?.metadata?.role ?? 'unknown',
                        content: String(doc?.pageContent ?? ''),
                        relevance_score: Number(doc?.metadata?.bm25Score ?? 0),
                    }));
            }
        }

        const fallbackScored = [];
        messages.forEach((msg, idx) => {
            const content = String(msg?.content ?? '');
            if (!content) return;
            const score = this._overlapScore(queryTokens, content);
            const overlapRatio = score / queryTokens.size;
            if (score >= 2 || overlapRatio >= 0.4) fallbackScored.push({ score, idx, msg });
        });

        return fallbackScored
            .sort((a, b) => (b.score - a.score) || (a.idx - b.idx))
            .slice(0, this.retrievalK)
            .sort((a, b) => a.idx - b.idx)
            .map(({ score, msg }) => ({
                role: msg?._getType?.() ?? msg?.getType?.() ?? msg?.type ?? msg?.role ?? 'unknown',
                content: String(msg?.content ?? ''),
                relevance_score: score,
            }));
    }

    async askQuestion(topic, provider = null, template = STRUCTURED_TEMPLATE, maxTokens = 1000, temperature = 0.7, sessionId = 'default') {
        const effectiveTemplate = template === '{topic}' ? STRUCTURED_TEMPLATE : template;
        const resolvedProvider = this.resolveModelIdentifier(provider);
        const basePrompt = effectiveTemplate.replace('{topic}', topic);

        if (!resolvedProvider) {
            return {
                success: false,
                error: 'No providers available',
                provider: 'none',
                model: 'none',
                prompt: basePrompt,
                response: null,
            };
        }

        const modelConfig = this.resolveModelConfig(resolvedProvider);

        let messages = [];
        if (this.retrievalMemoryEnabled) {
            const history = this._getHistory(sessionId);
            messages = await history.getMessages();
        }

        const retrieved = this.retrievalMemoryEnabled ? await this._selectRetrievedMessages(topic, messages) : [];
        const retrievedContext = retrieved.map((item) => `[${item.role}] ${item.content}`).join('\n');
        const retrievalAugmentedTopic = retrievedContext
            ? `Relevant memory snippets:\n${retrievedContext}\n\nCurrent user topic: ${topic}`
            : topic;
        const retrievalPrompt = effectiveTemplate.replace('{topic}', retrievalAugmentedTopic);

        const fullHistoryContext = messages
            .map((msg) => `[${msg?._getType?.() ?? msg?.getType?.() ?? msg?.type ?? 'unknown'}] ${String(msg?.content ?? '')}`)
            .join('\n');
        const promptWithoutRetrieval = fullHistoryContext
            ? `${fullHistoryContext}\n\nCurrent user topic: ${topic}`
            : topic;

        const tokensWithRetrieval = this._estimateTokens(retrievalPrompt);
        const tokensWithoutRetrieval = this._estimateTokens(promptWithoutRetrieval);
        const estimatedSaved = Math.max(0, tokensWithoutRetrieval - tokensWithRetrieval);
        const reductionPercent = tokensWithoutRetrieval > 0
            ? Number(((estimatedSaved / tokensWithoutRetrieval) * 100).toFixed(2))
            : 0;

        try {
            const model = this._createModel(resolvedProvider, temperature, maxTokens);
            const result = await model.invoke(this._buildMessages(retrievalPrompt));
            const rawResponse = this._extractText(resolvedProvider, result);

            const responseMetadata = result?.response_metadata ?? result?.responseMetadata ?? null;
            const usageMetadata = result?.usage_metadata ?? result?.usageMetadata ?? null;
            const tokenUsage = this._extractTokenUsage(responseMetadata, usageMetadata);

            const metadataPayload = {
                provider: modelConfig.provider,
                model: modelConfig.model,
                modelIdentifier: modelConfig.name,
                sessionId,
                temperature,
                maxTokens,
                response_metadata: responseMetadata,
                usage_metadata: usageMetadata,
            };

            let parsed;
            try {
                parsed = parseStructuredJsonResponse(rawResponse);
            } catch {
                parsed = {
                    answer: rawResponse,
                    summary: 'Model returned plain-text output instead of strict JSON.',
                    keywords: [],
                    distilled: rawResponse,
                    metadata: {
                        confidence: 'low',
                        notes: 'Structured parser fallback applied for non-JSON response.',
                    },
                };
            }
            parsed.metadata = {
                ...(parsed.metadata || {}),
                ...this._buildMetadata(metadataPayload, rawResponse),
                retrieval: {
                    history_messages_available: messages.length,
                    retrieved_messages_count: retrieved.length,
                    retrieved_messages: retrieved,
                    tokens_with_memory_retrieval: tokensWithRetrieval,
                    tokens_without_memory_retrieval: tokensWithoutRetrieval,
                    estimated_tokens_saved: estimatedSaved,
                    estimated_token_reduction_percent: reductionPercent,
                },
            };

            if (this.retrievalMemoryEnabled) {
                const history = this._getHistory(sessionId);
                await history.addUserMessage(topic);
                await history.addAIMessage(rawResponse);
            }

            return {
                success: true,
                provider: modelConfig.provider,
                model: modelConfig.model,
                modelIdentifier: modelConfig.name,
                prompt: retrievalPrompt,
                response: parsed,
                rawAnswer: parsed.answer ?? rawResponse,
                temperature,
                maxTokens,
                sessionId,
                ...(responseMetadata != null ? { response_metadata: responseMetadata } : {}),
                ...(usageMetadata != null ? { usage_metadata: usageMetadata } : {}),
                ...(tokenUsage != null ? { token_usage: tokenUsage } : {}),
            };
        } catch (error) {
            return {
                success: false,
                provider: modelConfig.provider,
                model: modelConfig.model,
                modelIdentifier: modelConfig.name,
                prompt: retrievalPrompt,
                error: error.message,
                response: null,
                temperature,
                maxTokens,
                sessionId,
            };
        }
    }
}

async function main() {
    const args = process.argv.slice(2);
    if (args.includes('-h') || args.includes('--help')) {
        printCliHelp(process.argv[1]);
        return;
    }
    if (args.includes('web')) {
        const { runWebServer } = await import('../../../shared/essentials/web.ts');
        await runWebServer(() => new LangChainLLMManager(true));
    } else {
        const manager = new LangChainLLMManager(true);
        await manager._checkProviders();
        await interactiveCli(manager);
    }
}

export { LangChainLLMManager };

if (import.meta.url === `file://${process.argv[1]}`) {
    main().catch(console.error);
}
